flink-kubernetes-operator-autoscaler/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/AutoscalerFlinkMetrics.java

```diff
@@ -62,54 +61,49 @@ public AutoscalerFlinkMetrics(MetricGroup metricGroup) {
-                            var jobVertexMg =
-                                    metricGroup.addGroup(JOB_VERTEX_ID, jobVertexID.toHexString());
-
-                            evaluated.forEach(
-                                    (sm, esm) -> {
-                                        var smGroup = jobVertexMg.addGroup(sm.name());
-
-                                        smGroup.gauge(
-                                                CURRENT,
-                                                () ->
-                                                        Optional.ofNullable(
-                                                                        currentVertexMetrics.get())
-                                                                .map(m -> m.get(jobVertexID))
-                                                                .map(metrics -> metrics.get(sm))
-                                                                .map(
-                                                                        EvaluatedScalingMetric
-                                                                                ::getCurrent)
-                                                                .orElse(Double.NaN));
-
-                                        if (sm.isCalculateAverage()) {
-                                            smGroup.gauge(
-                                                    AVERAGE,
-                                                    () ->
-                                                            Optional.ofNullable(
-                                                                            currentVertexMetrics
-                                                                                    .get())
-                                                                    .map(m -> m.get(jobVertexID))
-                                                                    .map(metrics -> metrics.get(sm))
-                                                                    .map(
-                                                                            EvaluatedScalingMetric
-                                                                                    ::getAverage)
-                                                                    .orElse(Double.NaN));
-                                        }
-                                    });
-                        });
+                    metricsSupplier) {
+        if (scalingMetricsInitialized) {
+            // It is important that we only initialize these metrics once because the Flink API does
+            // not support registering counters / gauges multiple times.
+            // The metrics will be updated via the provided metricsSupplier.
+            return;
+        }
+        scalingMetricsInitialized = true;
+
+        LOG.info("Registering scaling metrics");
+        for (JobVertexID jobVertexID : jobVerticesSupplier.get()) {
```

Comment:
The `jobVerticesSupplier.get()` is just called once, and it's not the lazy call. so the `jobVerticesSupplier` parameter can be simplified to `jobVertices`, right?

Reply:
Getting the job vertices is an expensive operation. We want to avoid calling it on every call to this method which will happen for every reconciliation loop.

Reply:
Thanks for the clarification, the optimization makes sense to me.

Could you help adding a comment here? This optimization is implicit, and adding comments can prevent other developers from removing this optimization during refactoring in the future.

Reply:
Good point. I've refactored the code, to cache the topological ordering because we call this at multiple places already. That way we can remove the Supplier for the vertices.